File: AI LAB/04_CSP/CSP.cpp
// Time Complexity	O(k^n)	
// 𝑘
// k colors for 
// 𝑛
// n vertices, exploring all combinations.
// Space Complexity	O(n)	Space for the colors array and recursion stack.


#include <iostream>
#include <vector>
#include <cstring>

class CSP {
private:
    std::vector<std::vector<int>> graph;
    int numColors;
    std::vector<int> colors;

public:
    CSP(const std::vector<std::vector<int>>& graph, int numColors)
        : graph(graph), numColors(numColors), colors(graph.size(), -1) {}

    bool isSafe(int vertex, int color) {
        for (size_t i = 0; i < colors.size(); i++) {
            if (graph[vertex][i] == 1 && colors[i] == color) {
                return false;
            }
        }
        return true;
    }

    bool solve(int vertex) {
        if (vertex == graph.size()) return true;

        for (int i = 0; i < numColors; i++) {
            if (isSafe(vertex, i)) {
                colors[vertex] = i;

                if (solve(vertex + 1)) return true;
                colors[vertex] = -1;  // Backtrack
            }
        }
        return false;
    }

    bool colorGraph() {
        return solve(0);
    }

    void printColors() {
        for (int color : colors) {
            std::cout << color << " ";
        }
        std::cout << std::endl;
    }
};

int main() {
    std::vector<std::vector<int>> graph = {
        {0, 1, 1, 1},
        {1, 0, 0, 1},
        {1, 0, 0, 1},
        {1, 1, 1, 0}
    };
    int numColors = 3;

    CSP csp(graph, numColors);
    if (csp.colorGraph()) {
        std::cout << "Coloring: ";
        csp.printColors();
    } else {
        std::cout << "No solution exists" << std::endl;
    }

    return 0;
}

File: AI LAB/04_CSP/CSP.java
import java.util.*;


public class CSP {
    private int[][] graph;
    private int numColors;
    private int[] colors;

    public CSP(int[][] graph,int numColors){
        this.graph=graph;
        this.numColors=numColors;
        this.colors=new int[graph.length];
        Arrays.fill(colors,-1);
    }

    private boolean isSafe(int vertex,int color){
        for (int i = 0; i < colors.length; i++) {
            if(graph[vertex][i]==1 && colors[i]==color){
                return false;
            }
        }
        return true;
    }


    public boolean solve(int vertex){
        if(vertex==graph.length) return true;

        for(int i=0;i<numColors;i++){
            if(isSafe(vertex, i)){
                colors[vertex]=i;

                if(solve(vertex+1)) return true;
                colors[vertex]=-1;
            }
        }
        return false;
    }

    public boolean colorGraph(){
        return solve(0);
    }


    public void printColors(){
        System.out.println("");
    }
    public static void main(String[] args) {
        int[][] graph={
            {0,1,1,1},
            {1,0,0,1},
            {1,0,0,1},
            {1,1,1,0}
        };
        int numColors=3;

        CSP csp=new CSP(graph, numColors);
        if(csp.colorGraph()){
            csp.colorGraph();
        }
        else{
            System.out.println("No solution Exist");
        }
    }
}